### DISCUSSION TITLE
An issue with autoload=yes in uosc.conf

### DISCUSSION THREAD

**Discussion by Raza-0007** [tomasklaen/uosc#1003] [General]
I have encountered an issue when I set `autoload=yes` in `uosc.conf`

with `keep-open=yes` and `save-position-on-quit`  set and a `watch-later-directory` specified in `mpv.conf`

If I play a folder with 4 video files, uosc will automatically moves to the next file, and mpv will remember each file's last position in the watch later directory, until mpv stops at the end of the last video file.

Everything works fine but the problem occurs if I relaunch the first video file again. mpv loads the file, with the position remembered i.e. the very end. This triggers uosc to load the next file, which also loads with the play position at the end, so file 3 is loaded and so on until everything stops at the end of file 4. This happens very quickly in a blink of an eye. I am unable to replay files 1-3 without deleting their saved file setting in the watch later directory. 

Is there a setting that I am missing? How do I prevent this from happening?

This issue does not happen with the new `--autocreate-playlist` in mpv 0.39, if used without uosc, so I am guessing something is missing in uosc that triggers this. 

Thanks,

~Raza

**Comment by tomasklaen** [comment]
It's two systems fighting each other. We implement `autoload` by setting up a timer that triggers shortly before file ends, and inside it we tell mpv to load a new file. That's because there's no mpv "on file end" event that would allow us to take over decision on which file to load next, so we have to simulate it his way.

And from mpv's perspective, the new file is loaded before the previous one ended, so it saves watch-later data for it.

Not sure what to do about this. We should probably just deprecate `autoload` in favor of `autocreate-playlist`

**Reply by hooke007** [reply]
Uosc autoload should be preserved.
Not everyone would like creating the playlist.

**Reply by Raza-0007** [reply]
@tomasklaen, first of all thank you very much for uosc. It is very fast and responsive, and contains just the right amount of options needed for a media player gui without all the bloat. It is a pleasure using mpv with it.

After reading your explanation about how uosc's `autoload` works, it makes sense why this happens. I was wondering if it is possible to implement a check in `autoload` where when a file is launched, `autolaod` checks the file's last saved position in the watch later directory, and if the position is at or a couple of seconds before the end of file, then instead of loading the next file, `autoload` just replays the same file again. This would resolve the issue. 

Its just an idea. I don't know if it is practical to implement it. 

In the mean time I can avoid this issue by preventing mpv from reaching the very end of file and just manually skip to the next file. 

~Raza

**Reply by Raza-0007** [reply]
@hooke007, I agree with you that `autoload` should not be depreciated. But correct me if I am wrong, isn't mpv's `autocreate-playlist` option doing the exact same thing as uosc's `autoload`? 

When you launch a file inside a directory, uosc auto creates a playlist of all the other video, audio, image files in that directory. `autoload` just loads the next file in that playlist. 

~Raza

**Reply by hooke007** [reply]
No, autoload doesn't create the playlist.

**Reply by Raza-0007** [reply]
After testing mpv's `autocreate-playlist` I found out that `autocreate-playlist` [does not save resume playback data](https://github.com/mpv-player/mpv/blob/master/player/loadfile.c#L1682-L1684), so all files always start from the beginning. This has its uses, for example, if one was creating a playlist of music files to listen, one most likely would not want files to start from the last played position. But in case someone is watching a multi-episode TV series, and does not remember the last episode played, the watch-later history is extremely important to determine where one let off last time.

uosc with `autoload` saves watch later history and this is what I prefer. If the issue I described earlier is resolved then in my opinion uosc's `autoload` is better then mpv's `autocreate-playlist` option in vast majority of use cases.

~Raza

**Reply by natural-harmonia-gropius** [reply]
> `autocreate-playlist` [does not save resume playback data](https://github.com/mpv-player/mpv/blob/master/player/loadfile.c#L1682-L1684)

Can you try this patch with `script-binding uosc/prev` and `script-binding uosc/next`? https://github.com/natural-harmonia-gropius/uosc/commit/c18e948f6da36d0ac6ed0a7101e3ad56a8350ed4

**Reply by Raza-0007** [reply]
@natural-harmonia-gropius,

I tested this with `autoload=no` in `uosc.conf` and `autocreate-playlist=filter` in `mpv.conf`

Everything seems to be working fine more or less! Episodes that have been played till the end, restart from beginning when played next time, and that domino effect that I described earlier does not happen. Files also retain their last played position. I also tested with uosc next and previous keybindings, and they work fine as well. 

The only issue I discovered is when you switch episodes from uosc's playlist button using a mouse, more then once, the first time the position is remembered, but the next time that same file starts from beginning. I will continue testing this further and let you know if I encounter any other issues. 

Thanks,

~Raza

**Reply by hooke007** [reply]
> autocreate-playlist [does not save resume playback data](https://github.com/mpv-player/mpv/blob/master/player/loadfile.c#L1682-L1684)

mpv doesn't save the 'position' in the playlist. It's not bug. For more info just check mpv's issue tracker. Example https://github.com/mpv-player/mpv/issues/9586#issuecomment-1375122257

https://mpv.io/manual/master/#command-interface-write-watch-later-config It's the native command.
Just bind it together with next command in your own input conf.

> https://github.com/natural-harmonia-gropius/uosc/commit/c18e948f6da36d0ac6ed0a7101e3ad56a8350ed4

I don't think always try to save postion in the playlist is a good idea for other use cases.(i.e. Playing all the music in the the directory, next time when you play the music it would alway tried to skip to the timepos the last time saved.)

**Reply by Raza-0007** [reply]
> https://mpv.io/manual/master/#command-interface-write-watch-later-config It's the native command. Just bind it together with next command in your own input conf.
> 

@hooke007, thanks for the info. I was however unable to get it to work. I believe it is due to "delete watch-later data" hard coded in the `autocreate-playlist` command. So the file's watch-later data is saved when you move away from the file to the next file, but every time you move back to the same file, `autocreate-playlist` deletes that file's watch later data before playing the file.

I have resolved my issue by keeping `autoload=no` in `uosc.conf`. This prevents the next track from being loaded automatically, so the domino effect I described in my first post does not happen when I relaunch an already played file. And I can always move back and forth between files using uosc's autoloaded playlist manually. So my issue is more or less resolved.

Thanks,

~Raza
